### ISSUE TITLE
Missing information in Immunisation Web Template document

### ISSUE THREAD

**Issue by robtweed** [ethercis/ethercis#113]
I'm going to add this as a new issue, although it relates to #74 

I'm looking at the Immunisation template: IDCR - Immunisation summary.v0

I was expecting to find "time" as one of the fields at the same level as "comment" and "reason_witheld", but it's missing:

              {
                "min": 0,
                "aql_path": "/content[openEHR-EHR-ACTION.immunisation_procedure.v1]/description[at0001]/items[at0015]",
                "max": 1,
                "name": "Comment",
                "description": "An additional comment about the vaccination process.",
                "id": "comment",
                "category": "ELEMENT",
                "type": "DV_TEXT",
                "ethercis_sql": "\"ehr\".\"entry\".\"entry\" #>> '{/content[openEHR-EHR-ACTION.immunisation_procedure.v1],0,/description[at0001],/items[at0015]}'",
                "constraints": [
                  {
                    "mandatory_attributes": [
                      {
                        "attribute": "value",
                        "type": "STRING"
                      }
                    ],
                    "attribute_name": "value",
                    "constraint": {
                      "occurrence": {
                        "min": 1,
                        "max_op": "<=",
                        "min_op": ">=",
                        "max": 1
                      }
                    },
                    "type": "DV_TEXT"
                  }
                ],
                "node_id": "at0015"
              },
              {
                "min": 0,
                "aql_path": "/content[openEHR-EHR-ACTION.immunisation_procedure.v1]/description[at0001]/items[at0005]",
                "max": 1,
                "name": "Reason withheld",
                "description": "The reason why a vaccination was withheld.",
                "id": "reason_withheld",
                "category": "ELEMENT",
                "type": "DV_TEXT",
                "ethercis_sql": "\"ehr\".\"entry\".\"entry\" #>> '{/content[openEHR-EHR-ACTION.immunisation_procedure.v1],0,/description[at0001],/items[at0005]}'",
                "constraints": [
                  {
                    "mandatory_attributes": [
                      {
                        "attribute": "value",
                        "type": "STRING"
                      }
                    ],
                    "attribute_name": "value",
                    "constraint": {
                      "occurrence": {
                        "min": 1,
                        "max_op": "<=",
                        "min_op": ">=",
                        "max": 1
                      }
                    },
                    "type": "DV_TEXT"
                  }
                ],
                "node_id": "at0005"
              }

The Marand Web Template includes it:

            {
              "id": "comment",
              "name": "Comment",
              "localizedName": "Comment",
              "rmType": "DV_TEXT",
              "nodeId": "at0015",
              "min": 0,
              "max": 1,
              "localizedNames": {
                "en": "Comment"
              },
              "localizedDescriptions": {
                "en": "An additional comment about the vaccination process."
              },
              "aqlPath": "/content[openEHR-EHR-ACTION.immunisation_procedure.v1]/description[at0001]/items[at0015]/value",
              "inputs": [
                {
                  "type": "TEXT"
                }
              ]
            },
            {
              "id": "reason_withheld",
              "name": "Reason withheld",
              "localizedName": "Reason withheld",
              "rmType": "DV_TEXT",
              "nodeId": "at0005",
              "min": 0,
              "max": 1,
              "localizedNames": {
                "en": "Reason withheld"
              },
              "localizedDescriptions": {
                "en": "The reason why a vaccination was withheld."
              },
              "aqlPath": "/content[openEHR-EHR-ACTION.immunisation_procedure.v1]/description[at0001]/items[at0005]/value",
              "inputs": [
                {
                  "type": "TEXT"
                }
              ]
            },
            {
              "id": "time",
              "name": "Time",
              "rmType": "DV_DATE_TIME",
              "min": 1,
              "max": 1,
              "annotations": {
                "view:VisibleInView": "time"
              },
              "aqlPath": "/content[openEHR-EHR-ACTION.immunisation_procedure.v1]/time",
              "inputs": [
                {
                  "type": "DATETIME"
                }
              ],
              "inContext": true
            },

As it happens, the Marand version includes a whole bunch of other fields at this level, including "language", "encoding" and "subject", none of which are in the EtherCIS version either (although I don't need them for the purposes of Ripple)

**Comment by freshehr** [comment]
The issue here is that the Marand web template includes a number of attributes from the openEHR RM, that are not part of the template /archetypes or indeed the .opt. Marad are pulling in small set of the more clinically relevant RM attributes into the web template.

The time that you are missing is the ACTION.time attribute associated with ACTION archetypes. This is the time that the action (in this case vaccination) was performed, which may be different from the date that the composition was authored.

**Comment by chevalleyc** [comment]
In fact the leveling used by Marand is kind of flatten (aqlPath is the reference). Template instrospection in EtherCIS does preserve the tree structure and the data you mention (language, encoding, subject, time, encoding...) are indeed present and at the correct level in the tree (e.g. at archetype level, here an ACTION).

**Comment by robtweed** [comment]
So does this mean that for Vaccinations, we can't use the automated "jumper" concept?  The time value is one of the fields we need for the UI - see:

https://github.com/RippleOSI/Ripple-openEHR/blob/master/docs/bindings/Current%20Ripple%20Headings/Vaccinations/immunisations.md

**Comment by robtweed** [comment]
There are a couple of problems with that "time" definition:

1) It doesn't have an "id" property, so my processing ignores it
2) It's in the wrong part of the JSON tree hierarchy to correctly construct the Flat JSON metadata.  I'd expect it to be in the same section as "comment" and "series_number", based on the definition provided by Ian:

"immunisation_summary/immunisation_procedure:0/immunisation_name": "Hepatitis B",
"immunisation_summary/immunisation_procedure:0/series_number": 3,
"immunisation_summary/immunisation_procedure:0/comment": "Immunisation complete",
"immunisation_summary/immunisation_procedure:0/time": "2016-12-20T07:45:20.622Z" //Vaccination time

eg: here's the "comment" definition which is fine:

              {
                "min": 0,
                "aql_path": "/content[openEHR-EHR-ACTION.immunisation_procedure.v1]/description[at0001]/items[at0015]",
                "max": 1,
                "name": "Comment",
                "description": "An additional comment about the vaccination process.",
                "id": "comment",
                "category": "ELEMENT",
                "type": "DV_TEXT",
                "ethercis_sql": "\"ehr\".\"entry\".\"entry\" #>> '{/content[openEHR-EHR-ACTION.immunisation_procedure.v1],0,/description[at0001],/items[at0015]}'",
                "constraints": [
                  {
                    "mandatory_attributes": [
                      {
                        "attribute": "value",
                        "type": "STRING"
                      }
                    ],
                    "attribute_name": "value",
                    "constraint": {
                      "occurrence": {
                        "min": 1,
                        "max_op": "<=",
                        "min_op": ">=",
                        "max": 1
                      }
                    },
                    "type": "DV_TEXT"
                  }
                ],
                "node_id": "at0015"
              },

I'd have expected the "time" definition to look similar to this

**Comment by chevalleyc** [comment]
I see for the missing 'id'. I will add that into the definition structure.

Re the 'time' location in the tree, for this template, 'time' is an attribute of ACTION materialized by openEHR-EHR-ACTION.immunisation_procedure.v1. Unless there is another time into this template (and could not find any), it cannot be locate at the same level as 'comment', 'reason', 'series...' which are all contained in the description of that ACTION. 

The Marand WebTemplate corroborate this as the definition is:

```
        {
          "id": "time",
          "name": "Time",
          "rmType": "DV_DATE_TIME",
          "min": 1,
          "max": 1,
          "annotations": {
            "view:VisibleInView": "time"
          },
          "aqlPath": "/content[openEHR-EHR-ACTION.immunisation_procedure.v1]/time",
          "inputs": [
            {
              "type": "DATETIME"
            }
          ],
          "inContext": true
        },
```
As you can see, 'time' is indeed an attribute of ACTION.

**Comment by robtweed** [comment]
After you add the id, I'll have a look and see if / how the required Flat JSON can be figured out from the template.
Thanks Christian

**Comment by freshehr** [comment]
@chevalleyc thanks I see it now!! I agree that you have placed time in the formally correct place according to the RM. Marand have deliberately flattened the RM structure is some places to try to make the web templates and therefore FLAT/STRCTURED JSON easier for devs not familiar with the RM. I can see arguments both ways, indeed we probably need both.

**Comment by robtweed** [comment]
OK that has sorted out the Web Template / Introspection document, so I'm now creating the correct structures for the JSON documents I derive from it (eg for Flat JSON construction).

However, there now appears to be a problem with the "top-level AQL" response, with time appearing in a very weird place, eg as here:

  "resultSet": [
    {
      "data": {
        "@class": "COMPOSITION",
        "content": [
          {
            "archetype_node_id": "openEHR-EHR-ACTION.immunisation_procedure.v1",
            "name": {
              "value": "Immunisation summary"
            },
            "time": {     <================================
              "name": {
                "value": "Immunisation procedure"
              },
              "value": {
                "value": "2017-11-02T22:00:00.000Z",
                "epoch_offset": 1509660000000,
                "@class": "DV_DATE_TIME"
              },
              "@class": "ELEMENT"
            },

Based on the introspection document info, I'm expecting it to be under the "immunisation_name" property, along with "series_number" and "comment", from which I derive this "reference" structure:

  "immunisation_procedure": {
    "immunisation_name": {
      "value": "{{immunisation_name}}",
      "code": "{{immunisation_name_codeString}}",
      "terminology": "{{immunisation_name_terminology}}"
    },
    "series_number": {
      "value": "{{series_number}}"
    },
    "comment": {
      "value": "{{comment}}",
      "code": "{{comment_codeString}}",
      "terminology": "{{comment_terminology}}"
    },
    "time": {
      "value": "{{time}}"
    }, ...etc

So the series_number and comment data are correctly placed in the "top-level AQL" response to allow me to create the above structure for those fields:

        "content": [
          {
            "archetype_node_id": "openEHR-EHR-ACTION.immunisation_procedure.v1",
            "name": {
              "value": "Immunisation summary"
            },
            "time": {
              "name": {
                "value": "Immunisation procedure"
              },
              "value": {
                "value": "2017-11-02T22:00:00.000Z",
                "epoch_offset": 1509660000000,
                "@class": "DV_DATE_TIME"
              },
              "@class": "ELEMENT"
            },
            "description": {
              "archetype_node_id": "at0001",
              "items": [
                {
                  "name": {
                    "value": "Immunisation name"
                  },
                  "value": {
                    "value": "111111111111111",
                    "@class": "DV_TEXT"
                  },
                  "@class": "ELEMENT",
                  "archetype_node_id": "at0002"
                },
                {
                  "name": {
                    "value": "Series number"
                  },
                  "value": {
                    "accuracy": 0,
                    "magnitude": 1,
                    "accuracy_percent": false,
                    "@class": "DV_COUNT"
                  },
                  "@class": "ELEMENT",
                  "archetype_node_id": "at0004"
                },
                {
                  "name": {
                    "value": "Comment"
                  },
                  "value": {
                    "value": "fssdfsfs",
                    "@class": "DV_TEXT"
                  },
                  "@class": "ELEMENT",
                  "archetype_node_id": "at0015"
                }
              ]

**Comment by chevalleyc** [comment]
Fixed in v1.1.2
